eval/narrow.elv

```diff
@@ -1,78 +1,124 @@
+before-location = []
+after-location = []
+before-history = []
+after-history = []
+before-lastcmd = []
+after-lastcmd = []
+
 fn location {
-    candidates = [(dir-history | each {
-	score = (splits $0[score] &sep=. | take 1)
+  for hook $before-location { $hook }
+  candidates = [(dir-history | each [arg]{
+	      score = (splits . $arg[score] | take 1)
         put [
-            &content=$0[path]
-            &display=$score" "$0[path]
-	]
-    })]
+          &content=$arg[path]
+          &display=$score" "$arg[path]
+	      ]
+  })]
 
-    edit:-narrow-read {
-        put $@candidates
-    } {
-        cd $0[content]
-    } &modeline="Location "
+  edit:-narrow-read {
+    put $@candidates
+  } [arg]{
+    cd $arg[content]
+    for hook $after-location { $hook }
+  } &modeline="[narrow] Location " &ignore-case=$true
 }
 
 fn history {
-    candidates = [(edit:command-history | each {
+  for hook $before-history { $hook }
+  candidates = [(edit:command-history | each [arg]{
         put [
-	    &content=$0[cmd]
-	    &display=$0[id]" "$0[cmd]
+	        &content=$arg[cmd]
+	        &display=$arg[id]" "(replaces "\t" " " (replaces "\n" " " $arg[cmd]))
         ]
-    })]
+  })]
 
-    edit:-narrow-read {
-        put $@candidates
-    } {
-        edit:replace-input $0[content]
-    } &modeline="History " &keep-bottom=$true
+  edit:-narrow-read {
+    put $@candidates
+  } [arg]{
+    edit:replace-input $arg[content]
+    for hook $after-history { $hook }
+  } &modeline="[narrow] History " &keep-bottom=$true &ignore-case=$true
 }
 
 fn lastcmd {
-    last = (edit:command-history -1)
-    cmd = [
-            &content=$last[cmd]
-            &display="-1 "$last[cmd]
-	    &filter-text=""
-        ]
-    index = 0
-    candidates = [$cmd ( edit:wordify $last[cmd] | each {
-	put [
-            &content=$0
-            &display=$index" "$0
-            &filter-text=$index
-	]
-	index = (+ $index 1)
-    })]
-    edit:-narrow-read {
-        put $@candidates
-    } {
-        edit:replace-input $0[content]
-    } &modeline="Lastcmd " &auto-commit=$true &bindings=[&M-1={ edit:narrow:accept-close }]
+  for hook $before-lastcmd { $hook }
+  last = (edit:command-history -1)
+  cmd = [
+    &content=$last[cmd]
+    &display="M-1 "$last[cmd]
+	  &filter-text=""
+  ]
+  index = 0
+  candidates = [$cmd ( edit:wordify $last[cmd] | each [arg]{
+	      put [
+          &content=$arg
+          &display=$index" "$arg
+          &filter-text=$index
+	      ]
+	      index = (+ $index 1)
+  })]
+  edit:-narrow-read {
+    put $@candidates
+  } [arg]{
+    edit:insert-at-dot $arg[content]
+    for hook $after-lastcmd { $hook }
+  } &modeline="[narrow] Lastcmd " &auto-commit=$true &bindings=[&M-1={ edit:narrow:accept-close }] &ignore-case=$true
 }
 
 
-# TODO: separate bindings from functions
+fn -bind_i [k f]{
+  edit:insert:binding[$k] = $f
+}
 
-fn bind [m k f]{
-    edit:binding[$m][$k] = $f
+fn -bind_n [k f]{
```

Comment:
Consider calling this just `-bind`. When using this mode it will be `narrow:-bind`.